### PR TITLE
feat: rapid design system and web-core prelim FUI-1948

### PR DESCRIPTION
**Pull request by derekdon**
[Review Guidance](https://www.notion.so/genesisglobal/Platform-Code-Review-Process-ace93ba760cc4563b0dfb712e2a88d8a)

📓 &nbsp; **Related JIRA**

[FUI-1948](https://genesisglobal.atlassian.net/browse/FUI-1948)

🤔 &nbsp; **What does this PR do?**

- Adds prelim code to move to rapid and web core.
- Updates to latest to include newly released header that can support multiple Design Systems.

🚀 &nbsp; **Where should the reviewer start?**

Add file / directory pointers.

📑 &nbsp; **How should this be tested?**

### Defaults

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest -x --ref derekdon/rapid-web-core-prelim --no-npm
```

### Route and CSV parameter handling test

```
rm -rf blankappseedtest && npx -y @genesislcap/genx@latest init blankappseedtest --ref derekdon/rapid-web-core-prelim --no-npm -x --routes '[{"name":"Home","tiles":[{"title":"Entity manager","type":"entity-manager","config":{ "modalPosition": "centre", "sizeColumnsToFit": true, "enableSearchBar": true, "resourceName":"ALL_POSITIONS","title":"My Positions","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT", "createFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline create schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "updateFormUiSchema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline update schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]}, "columns": [{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}] }},{"title":"Grid","type":"grid-pro","config":{"resourceName":"ALL_TRADES", "deferredGridOptions": {"columnDefs":[{"field":"INSTRUMENT_NAME","headerName":"Instrument Name"},{"field":"VALUE","headerName":"Inline coldef - VALUE"},{"field":"QUANTITY","headerName":"Quantity"},{"field":"PNL","headerName":"PNL"}]} }},{"title":"Form","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT", "uischema": {"type":"VerticalLayout","elements":[{"type":"Control","label":"Inline form schema - main contact","scope":"#/properties/MAIN_CONTACT"},{"type":"Control","label":"Issuer Name - Local Schema","scope":"#/properties/ISSUER_NAME","options":{"readonly":true}},{"type":"Control","label":"Price","scope":"#/properties/PRICE"},{"type":"Control","scope":"#/properties/COUNTERPARTY","options":{"allOptionsResourceName":"ALL_COUNTERPARTYS","valueField":"COUNTERPARTY_ID","labelField":"COUNTERPARTY_ID"}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"isPassword":true}},{"type":"Control","label":"Password","scope":"#/properties/PASSWORD","options":{"textarea":true}}]} }}]}, {"name":"Realtime Dashboard","tiles":[{"title":"Entity manager tile","type":"entity-manager","config":{"resourceName":"ALL_COUNTERPARTYS","title":"Counterparty Management","updateEvent":"EVENT_COUNTERPARTY_MODIFY","deleteEvent":"EVENT_COUNTERPARTY_DELETE","createEvent":"EVENT_COUNTERPARTY_INSERT"}},{"title":"Form tile","type":"smart-form","config":{"resourceName":"EVENT_COUNTERPARTY_INSERT"}}]},{"name":"Analytics","tiles":[{"title":"Grid Tile","type":"grid-pro","config":{"resourceName":"ALL_POSITIONS"}},{"title":"Charts Tile 1","type":"chart","config":{"type":"line","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 2","type":"chart","config":{"type":"pie","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_NAME","yField":"VALUE"}},{"title":"Charts Tile 3","type":"chart","config":{"type":"column","resourceName":"ALL_POSITIONS","xField":"INSTRUMENT_ID","yField":"VALUE"}}]}]' --csv '[{"name": "trade", "fields": ["a", "B"]}, {"name": "position", "fields": ["id", "TYPE"]} ]' --apiHost 'wss://public-foundation.genesislab.global/gwf/' --no-shell && cd blankappseedtest/client && npm run bootstrap && npm run dev
```

✅ &nbsp; **Checklist**

- [ ] I have tested my changes.
- [ ] I have added tests for my changes.
- [ ] I have updated the project documentation to reflect my changes.

### CODE REVIEW

**Review by matteematt — COMMENTED** [review]
I will review properly tomorrow if someone else already hasn't, but typo here? https://github.com/genesiscommunitysuccess/blank-app-seed/pull/210/files#diff-bcdab84cfdae77caaa0f3befbc12a08ff9aef7ade30e7989ca64cc07420b47a5R1

File: .genx/templates/chart.hbs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-<zero-g2plot-chart
+@<rapid-g2plot-chart
```
Comment (cistov) [inline .genx/templates/chart.hbs:1]: `@`?

**Comment by cistov** [comment]
Looks like rapid package not published yet - build failing with:

```
npm ERR! 404 Not Found - GET https://registry.npmjs.org/@genesislcap%2ffoundation-rapid - Not found
```

**Comment by matteematt** [comment]
> Looks like rapid package not published yet - build failing with:
> 
> ```
> npm ERR! 404 Not Found - GET https://registry.npmjs.org/@genesislcap%2ffoundation-rapid - Not found
> ```

It's because the packages is `@genesislcap/rapid-design-system`

File: client/src/main/main.ts [hunk 2]
```diff
@@ -3,10 +3,18 @@ import { EventEmitter } from '@genesislcap/foundation-events';
...
+  customElement,
+  Container,
+  DefaultRouteRecognizer,
+  DOM,
+  GenesisElement,
+  inject,
+  observable,
+  Registration,
+  StandardLuminance
+} from '@genesislcap/web-core';
```
Comment (cistov) [inline client/src/main/main.ts:17]: Fyi, we run eslint/prettier on the generated app during CI, so any changes like this will need to pass

**Comment by derekdon** [comment]
bunch of hyenas... :-) pushed

File: .genx/templates/chart.hbs (hunk 1 above)
Comment (derekdon) [reply .genx/templates/chart.hbs]: Good catch... fixed

File: client/src/components/components.ts [hunk 3]
```diff
@@ -42,12 +42,12 @@ async function zeroDesignSystemImport(): Promise<DesignSystemModule> {
...
     module = await import(
-      /* webpackChunkName: "foundation-zero" */
-      '@genesislcap/foundation-zero'
+      /* webpackChunkName: "rapid-design-system" */
+      '@genesislcap/rapid-design-system'
     );
     return assureDesignSystem(module);
   } finally {
-    logger.debug(`Using '${type}' version of foundation-zero`);
+    logger.debug(`Using '${type}' version of rapid-design-system`);
```
Comment (cistov) [inline client/src/components/components.ts:50]: Not related to this PR: this boilerplate for attempting to load a federated DS - we keep adding it to all projects even though no one's loading their DSes via federation (as far as I know). Not sure whether it needs to be enabled by default or could be a page in the docs explaining how to enable it if you want to.

**Comment by derekdon** [comment]
> > Looks like rapid package not published yet - build failing with:
> > ```
> > npm ERR! 404 Not Found - GET https://registry.npmjs.org/@genesislcap%2ffoundation-rapid - Not found
> > ```
> 
> It's because the packages is `@genesislcap/rapid-design-system`

Yeah I'd misnamed the package... fixed

File: client/src/main/main.ts (hunk 2 above)
Comment (derekdon) [reply client/src/main/main.ts:17]: Okay cool...

File: client/src/components/components.ts (hunk 3 above)
Comment (derekdon) [reply client/src/components/components.ts]: Removed...

**Comment by derekdon** [comment]
Right it builds with Rapid, users can use... however, need to cascade to the header etc. Will look again tmw.

Screenshot 2024-05-16 at 21 45 27

**Comment by derekdon** [comment]
Please try this branch via the `Route and CSV parameter handling test` when you get time. It should help spot any ds issues.

**Comment by derekdon** [comment]
Putting back into draft for a little while given we now will be supporting both Zero and Rapid here... so want to ensure we haven't broken the ability to support Zero for any included PBCs that haven't moved over. Thanks @skawian for mentioning...

**Review by cistov — APPROVED** [review]

**Pull request merged**